iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs

```diff
@@ -0,0 +1,402 @@
+
+        // Dictionary to hold the max temperature since last reboot, for each "Thermostat" component.
+        private readonly Dictionary<string, double> _maxTemp = new Dictionary<string, double>();
+
+        public TemperatureControllerSample(DeviceClient deviceClient, ILogger logger)
+        {
+            _deviceClient = deviceClient ?? throw new ArgumentNullException($"{nameof(deviceClient)} cannot be null.");
+            _logger = logger ?? LoggerFactory.Create(builer => builer.AddConsole()).CreateLogger<TemperatureControllerSample>();
+        }
+
+        public async Task PerformOperationsAsync(CancellationToken cancellationToken)
+        {
+            // This sample follows the following workflow:
+            // -> Set handler to receive "reboot" command - root interface.
+            // -> Set handler to receive "getMaxMinReport" command - on "Thermostat" components.
+            // -> Set handler to receive "targetTemperature" property updates from service - on "Thermostat" components.
+            // -> Update device information on "deviceInformation" component.
+            // -> Send initial device info - "workingSet" over telemetry, "serialNumber" over reported property update - root interface.
+            // -> Periodically send "temperature" over telemetry - on "Thermostat" components.
+            // -> Send "maxTempSinceLastReboot" over property update, when a new max temperature is set - on "Thermostat" components.
+
+            _logger.LogDebug($"Set handler for \"reboot\" command.");
+            await _deviceClient.SetMethodHandlerAsync("reboot", HandleRebootCommandAsync, _deviceClient, cancellationToken);
+
+            // For a component-level command, the command name is in the format "<component-name>*<command-name>".
+            _logger.LogDebug($"Set handler for \"getMaxMinReport\" command.");
+            await _deviceClient.SetMethodHandlerAsync("thermostat1*getMaxMinReport", HandleMaxMinReportCommandAsync, Thermostat1, cancellationToken);
+            await _deviceClient.SetMethodHandlerAsync("thermostat2*getMaxMinReport", HandleMaxMinReportCommandAsync, Thermostat2, cancellationToken);
+
+            _logger.LogDebug($"Set handler to receive \"targetTemperature\" updates.");
+            await _deviceClient.SetDesiredPropertyUpdateCallbackAsync(SetDesiredPropertyUpdateCallbackAsync, null, cancellationToken);
+            _desiredPropertyUpdateCallbacks.Add(Thermostat1, TargetTemperatureUpdateCallbackAsync);
+            _desiredPropertyUpdateCallbacks.Add(Thermostat2, TargetTemperatureUpdateCallbackAsync);
+
+            await UpdateDeviceInformationAsync(cancellationToken);
+            await SendDeviceMemoryAsync(cancellationToken);
+            await SendDeviceSerialNumberAsync(cancellationToken);
+
+            bool temperatureReset = true;
+            _maxTemp[Thermostat1] = 0d;
+            _maxTemp[Thermostat2] = 0d;
+
+            await Task.Run(async () =>
```

Comment:
The context of this method is async Task, I don't think you need Task.Run here do you?